oteapi/plugins/entry_points.py

```diff
@@ -0,0 +1,505 @@
+"""Load plugins through entry points.
+
+This module deals with handling all plugged in strategies through the entry points API
+and importlib metadata API.
+
+Special functionality is put in place to handle the OTE-API-specific entry points.
+
+Since the entry points are information complete in and of themselves, there is no need
+to import actual strategy classes until they are truly needed.
+This therefore implements lazy loading of all plugin strategies.
+"""
+import importlib
+import re
+from collections import abc
+from enum import Enum
+from functools import lru_cache
+from importlib.metadata import entry_points as get_entry_points
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:  # pragma: no cover
+    from importlib.metadata import EntryPoint
+    from typing import Any, Iterable, Iterator, Optional, Set, Tuple, Type, Union
+
+    from oteapi.interfaces import IStrategy
+
+
+class EntryPointNotFound(Exception):
+    """A given strategy implementation (class) cannot be found from the given entry
+    point value."""
+
+
+class StrategyType(Enum):
+    """An enumeration of available strategy types.
+
+    Available strategy types:
+
+    - download
+    - filter
+    - mapping
+    - parse
+    - resource
+    - transformation
+
+    """
+
+    DOWNLOAD = "download"
+    FILTER = "filter"
+    MAPPING = "mapping"
+    PARSE = "parse"
+    RESOURCE = "resource"
+    TRANSFORMATION = "transformation"
+
+    @lru_cache
```

Comment:
What is the purpose of this cache? The dict lookup below is really fast and shouldn't need any cache...

Reply:
I think my original idea here is to have it store the dictionary in the cache instead of creating it anew each time the method is called? But maybe that's irrelevant?

Reply:
And if I understand `lru_cache` correctly, then it merely caches the return value based on the input parameters - so eventually it will have cached the whole dictionary.

Reply:
In this case it is not a big deal. But the `lru_cache` is essentially a dict with some additional logics for eviction++. Since what it is caching is already a dict, it just adds an unnessary additional layer without any benefits.